Sources/BSON/BSONDocumentIterator.swift

```diff
@@ -1,32 +1,59 @@
 import Foundation
 import NIO
 
-extension Document: Sequence {
+extension BSONDocument: Sequence {
     // Since a `Document` is a recursive structure, we want to enforce the use of it as a subsequence of itself,
     // instead of something like `Slice<Document>`.
     /// The type that is returned from methods such as `dropFirst()` and `split()`.
-    public typealias SubSequence = Document
+    public typealias SubSequence = BSONDocument
 
     /// Returns a `Bool` indicating whether the document is empty.
-    public var isEmpty: Bool { fatalError("Unimplemented") }
+    public var isEmpty: Bool { self.keySet.isEmpty }
 
     /// Returns a `DocumentIterator` over the values in this `Document`.
-    public func makeIterator() -> DocumentIterator {
-        fatalError("Unimplemented")
+    public func makeIterator() -> BSONDocumentIterator {
+        BSONDocumentIterator(over: self.buffer)
     }
 }
 
-public struct DocumentIterator: IteratorProtocol {
+public struct BSONDocumentIterator: IteratorProtocol {
     /// The buffer we are iterating over.
     private var buffer: ByteBuffer
 
     internal init(over buffer: ByteBuffer) {
-        fatalError("Unimplemented")
+        self.buffer = buffer
+        // moves readerIndex to first key's type indicator
+        self.buffer.moveReaderIndex(to: 4)
     }
 
     /// Advances to the next element and returns it, or nil if no next element exists.
     public mutating func next() -> (String, BSON)? {
-        fatalError("Unimplemented")
+        guard self.buffer.readableBytes != 0 else {
+            // Iteration has been exhuasted
+            return nil
+        }
+
+        guard let typeByte = self.buffer.readInteger(as: UInt8.self) else {
+            fatalError("BSONDocumentIterator Failed: Cannot read from \(self.buffer)")
+        }
+
+        guard let type = BSONType(rawValue: typeByte), type != .invalid else {
+            if typeByte == 0 {
```

Comment:
I think it might be a bit clearer to first check if the typeByte is 0 and then return nil, and then handle the invalid case afterward